types/wlr_input_method_v2.c

```diff
@@ -101,10 +119,177 @@ static void im_get_input_popup_surface(struct wl_client *client,
 	wlr_log(WLR_INFO, "Stub: zwp_input_method_v2::get_input_popup_surface");
 }
 
+static void im_keyboard_grab_destroy(
+		struct wlr_input_method_keyboard_grab_v2 *im_keyboard_grab) {
+	im_keyboard_grab->input_method->im_keyboard_grab = NULL;
+	if (im_keyboard_grab->keyboard) {
+		wl_list_remove(&im_keyboard_grab->keymap_listener.link);
+		wl_list_remove(&im_keyboard_grab->repeat_info_listener.link);
+
+		// send modifier state to original client
+		wlr_seat_keyboard_notify_modifiers(im_keyboard_grab->input_method->seat,
```

Comment:
Hmm. I wonder if the compositor should be responsible for this.

Reply:
I've moved this to compositor side.